frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/PullBackgroundable.java

```diff
@@ -0,0 +1,211 @@
+import git4idea.util.LocalChangesWouldBeOverwrittenHelper;
+import lombok.CustomLog;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@CustomLog
+public class PullBackgroundable extends Task.Backgroundable {
+
+  private static final String OPERATION_NAME = "Pull";
+
+  private final Project project;
+  private final GitRepository gitRepository;
+  private final GitLineHandler handler;
+  private final String remoteBranchName;
+
+  public PullBackgroundable(
+      Project project,
+      GitRepository gitRepository,
+      GitLineHandler handler,
+      String remoteBranchName) {
+    super(project,
+        /* taskTitle */ GitMacheteBundle.message("action.GitMachete.BasePullBranchAction.task-title"),
+        /* canBeCancelled */ true);
+    this.project = project;
+    this.gitRepository = gitRepository;
+    this.handler = handler;
+    this.remoteBranchName = remoteBranchName;
+  }
+
+  @Override
+  public void run(ProgressIndicator indicator) {
+    var localChangesDetector = new GitLocalChangesWouldBeOverwrittenDetector(gitRepository.getRoot(), MERGE);
+    var untrackedFilesDetector = new GitUntrackedFilesOverwrittenByOperationDetector(gitRepository.getRoot());
+
+    handler.addLineListener(localChangesDetector);
+    handler.addLineListener(untrackedFilesDetector);
+
+    Label beforeLabel = LocalHistory.getInstance().putSystemLabel(project, /* name */ "Before update");
+
+    GitUpdatedRanges updatedRanges = deriveGitUpdatedRanges(project, gitRepository, remoteBranchName);
+
+    String beforeRevision = gitRepository.getCurrentRevision();
+    try (AccessToken ignore = DvcsUtil.workingTreeChangeStarted(project, OPERATION_NAME)) {
+      GitCommandResult result = Git.getInstance().runCommand(() -> handler);
+
+      if (beforeRevision != null) {
+        GitRevisionNumber currentRev = new GitRevisionNumber(beforeRevision);
+        handleResult(result,
+            project,
+            localChangesDetector,
+            untrackedFilesDetector,
+            gitRepository,
+            currentRev,
+            beforeLabel,
+            updatedRanges);
+      }
+    }
+  }
+
+  @Nullable
+  private static GitUpdatedRanges deriveGitUpdatedRanges(
+      Project project,
+      GitRepository gitRepository,
+      String remoteBranchName) {
+    GitUpdatedRanges updatedRanges = null;
+    var currentBranch = gitRepository.getCurrentBranch();
+    if (currentBranch != null) {
+      String selectedBranch = StringUtil.trimStart(remoteBranchName, /* prefix */ "remotes/");
+      GitBranch targetBranch = gitRepository.getBranches().findBranchByName(selectedBranch);
+      if (targetBranch != null) {
+        GitBranchPair refPair = new GitBranchPair(currentBranch, targetBranch);
+        updatedRanges = GitUpdatedRanges.calcInitialPositions(project,
+            java.util.Collections.singletonMap(gitRepository, refPair));
+      } else {
+        LOG.warn("Couldn't find the branch with name [" + selectedBranch + "]");
```

Comment:
Nit: I don't think we have `[...]` convention for branch names in logs... just `'...'`